### ISSUE TITLE
New version of cli `ng build --prod` gives error.

### ISSUE THREAD

**Issue by webcat12345** [angular/angular-cli#5733]
### Bug Report
```
- [X] bug report
- [ ] feature request
```

### Versions.
Output from: `ng --version`.

```
@angular/cli: 1.0.0-rc.2
node: 7.6.0
os: linux x64
@angular/cli: 1.0.0-rc.2
@angular/common: 2.4.10
@angular/compiler: 2.4.10
@angular/core: 2.4.10
@angular/forms: 2.4.10
@angular/http: 2.4.10
@angular/platform-browser: 2.4.10
@angular/platform-browser-dynamic: 2.4.10
@angular/router: 3.4.10
```
### Repro steps.
Nothing changed.
Problem is happening after upgrading *CLI version to rc.2.*

I read the issue https://github.com/angular/angular-cli/issues/5623 , but it is closed without any solution. It is nearly impossible to change all properties to *public*, because project is too big.

I tried to change, but I gave up now.

Strange thing is build on Github through *Codeship* is succeed. How is it possible? And why not on local?

### The log given by the failure.
```
ERROR in /home/rango/Desktop/test/accepter/ng-reason/src/$$_gendir/app/components/home/home.component.ngfactory.ts (1038,49): Property 'test_detail_mode' is private and only accessible within class 'HomeComponent'.
/home/rango/Desktop/test/accepter/ng-reason/src/$$_gendir/app/components/home/home.component.ngfactory.ts (1041,44): Property 'test_detail_mode' is private and only accessible within class 'HomeComponent'.

ERROR in /home/rango/Desktop/test/accepter/ng-reason/src/$$_gendir/app/components/reason/reason.component.ngfactory.ts (1230,44): Property 'hours' is private and only accessible within class 'ReasonComponent'.
/home/rango/Desktop/test/accepter/ng-reason/src/$$_gendir/app/components/reason/reason.component.ngfactory.ts (1234,50): Property 'statistic' is private and only accessible within class 'ReasonComponent'.
... ... ...
```

### Desired functionality.
Is there anyway to upgrade codebase without changing all issues by manually?
Migrating all issues line by line is too pain for us, also creating `getter, setter` is same for us.

### Mention any other details that might be useful.

I think if this error is too critical then CLI should gives error on *ng serve or ng build*. 

Thank you very much.

**Comment by TheZeroNine** [comment]
Hi,
I had the same issue. Follow this tips to get everything working again: https://github.com/qdouble/angular-webpack2-starter#aot--donts

**Comment by JosuaMeier** [comment]
Hi,
Thanks @TheZeroNine 
I had also the same issue. But it only appears on 'ng build --prod --aot' and not on 'ng serve'.
Is it possible to see these error before starting a build. (don't want to start build after every change -.-)

**Comment by doyevaristo** [comment]
I have these after upgrading form angular cli beta into version 1.0. 
the sad part is my project is too big to handle this :(

**Comment by ginkosen** [comment]
Hi, I had the same issue when running **ng build --aot --prod -sm false**.
`
TerminalListComponent.ngfactory.ts (10181,35): Supplied parameters do not match any signature of call target.
`
And I can't find the ngfactory.ts file ,so i don't know how to solve it . so sad ...

**Comment by webcat12345** [comment]
Check `TerminalListComponent`, then you can see this kind of issue.
```
HTML : 
<element (click)="customfn(a, b)"></element>

Typescript : 
public customfn(param:any):void {
...
}
```

As you can see Supplied parameters do not match any signature of call target.

**Comment by ophirbushi** [comment]
I also can't find the ngfactory files after build errors.

**Comment by dhilgarth** [comment]
I am having this issue as well. Is there a way to find that ngfactory file?
Also, it would be great if the error message could show the name of the call target or something similar to identify the offending method

**Comment by ginkosen** [comment]
@webcat12345 Thx , I check my code again , and i find the problem and slove.

**Comment by webcat12345** [comment]
Problem is some *npm packages* are having same issues. They should be upgraded to latest *angular* and *angular/cli* version since we are using *Angular4* and *angular/cli 1.0.0*.

**Comment by da45** [comment]
why ng serve and ng build (without --prod option) are OK ?!!!!!

**Comment by doyevaristo** [comment]
make the property hours,test_detail_mode,statistic public.

**Comment by xealot** [comment]
I am having issues with `ng build` and `ng serve` when AOT is enabled.

The errors are along the lines of:
`$$_gendir/node_modules/.../XYZ.ngfactory.ts (1,1): Cannot find module 'XYZ'.`

and

```
ERROR in ./src/$$_gendir/app/app.module.ngfactory.ts
Module not found: Error: Can't resolve 'XYZ' in '/.../demo/src/$$_gendir/app'
 @ ./src/$$_gendir/app/app.module.ngfactory.ts 24:0-37
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://localhost:3000 ./src/main.ts
```

Typescript compiles cleanly, I have looked at the ["don'ts" page](https://github.com/qdouble/angular-webpack2-starter#aot--donts). It also works perfectly in either build or serve with `--aot false`.

A lot of other people seem to be having similar issues:
https://github.com/angular/angular-cli/issues/6013
https://github.com/angular/angular-cli/issues/6020
https://github.com/angular/angular-cli/issues/5802
https://github.com/angular/angular-cli/issues/5765

If we could get a better error message, or have the option of saving the files somewhere to inspect maybe it would be a little less opaque.

**Comment by mensahGit** [comment]
I hope this is of help to someone out there, wondering why their recently updated angular4(...but only moments ago, angular2) project won't build without throwing this decipherable message
_"ERROR in /Users/me/Desktop/kodotest/src/$$_gendir/app/app.component.ngfactory.ts (4352,35): Supplied parameters do not match any signature of call target."_

Aha...ok...thanks for that! What ever could you mean?...First of all, I don't own a factory...ahem!
Ok, enough jibbering...DO THIS!(I'm on a Mac.Not sure if Terminal prompts are identical on Windows.)
1) In Terminal execute: "**npm install -g angular-cli-ghpages**"
2) In your app's 'package.json', add the line:  **"deploy": "ng build -sm -ec -bh /aboutme/ & ngh --silent=false",** 
3)Finally, in Terminal execute: "**npm run deploy**"

Strangely enough, I still get an error.....(error: unknown option `--silent' ) but IT STILL BUILDS!
And outputs my _dist_ folder...

Hope this saves someone an hour.

**Comment by dherges** [comment]
I love the CLI, guys!

Can you write the generated files in `$$_gendir` to disk, so that AoT users can debug?

**Comment by LeonardoBluesMan** [comment]
same problem with cli 1.0.0, but I downgrade to 1.0.0-rc4 and it works without errors.

**Comment by Michael-D-Pine** [comment]
So just wanted to add my 2cents (australian, not sure how much that is US).

We were struggling with these problems and it is I think not reasonable for ng serve to work and then have build issues but I probably just don't have enough fingers in the pie's of Angular/cli to know how it all hangs together, anyway.

We are now running everything at latest, 4.1.0 and CLI 1.0.1.

We updated all of our dependant packages of which thankfull we don't have many but some awesome people had already upgraded their projects to use Angular 4, some olders are still not updated but are working fine.

ng serve and prod build now all work fine, so I think we were stuck in a loop of cli version mismatching with angular version and also some 3rd party packages. I did go through and change custom filters we had created to ensure the right number of parameters were matching.

I know this could be a big step for some people but for us it was worth it and 2 of us sat down with a strong cup of tea and knocked it over in 2 hours.

Just got 1 package that a previous developer got from somewhere which has to be completely replaced.

**Comment by filipesilva** [comment]
I'm sorry but this issue really is a duplicate of https://github.com/angular/angular-cli/issues/5623. The fix for the outstanding errors really is to fix your templates (e.g. change the variables that are private and need to be accessed in templates to public).

As mentioned in #5623 these really are AOT errors and it was a CLI bug that they weren't reported before.

We agree that it's a very bad experience to non-AOT builds to work, but then for AOT builds to not work. The plan for the future is to have AOT builds be fast enough they can be used even during development so this doesn't happen anymore, but we're not there yet.

**Comment by dhilgarth** [comment]
I think the main problem here is not that the build breaks but the fact that we get no info about where it breaks. My error was in some package I was using. Tracking that down is just a big PITA.

**Comment by dherges** [comment]
#5601 gives better error reporting. We just need to be alittle patient for it to be released. Maybe a 1.0.2 ?

**Comment by MarvelMoe** [comment]
Not sure if this was mentioned, but I found  "private": true in package.json. After removing this line and restarting the app,  ng build worked for me

**Comment by Michael-D-Pine** [comment]
Me, personally that seems like "possibly" a good way to hide certain behaviour, without reading I don't know what that setting does, I would prefer to have everything up to date and building correctly which just took us a bit of time.

I have total faith in the team because they fixing things very well as they come up on subsequent minor releases and we are very happy with what is coming out.

**Comment by iansamz** [comment]
This link wasvery helpful
https://github.com/qdouble/angular-webpack2-starter#aot--donts
Still have a problem using pipes

**Comment by LeonardoBluesMan** [comment]
We have fixed all private methods and variables that were called from html
and now it builds perfect.

**Comment by ben-coble-sp** [comment]
@iansamz By pipes do you mean pipes used to transform data or do you mean pipes used to indicate that parameters may be of several types (aka | as an OR operator)? If it's the later, see this comment https://github.com/angular/angular-cli/issues/5623#issuecomment-289577009

**Comment by iansamz** [comment]
@ben-coble-sp fixed it just an error on my side

**Comment by apidesh** [comment]
I'm still error with this issue.


$ ng --version
    _                      _                 ____ _     ___
   / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
  / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
 / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
/_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
               |___/
@angular/cli: 1.1.3
node: 8.1.2
os: win32 x64
@angular/animations: 4.2.4
@angular/common: 4.2.4
@angular/compiler: 4.2.4
@angular/core: 4.2.4
@angular/forms: 4.2.4
@angular/http: 4.2.4
@angular/material: 2.0.0-beta.7
@angular/platform-browser: 4.2.4
@angular/platform-browser-dynamic: 4.2.4
@angular/platform-server: 4.2.4
@angular/router: 4.2.4
@angular/cli: 1.1.3
@angular/compiler-cli: 4.2.4

**Comment by drjackson** [comment]
I had an error in my templates which only showed up when building with --aot.  I Isolated it by removing template to make sure that's where it was and then just searching through the events until I found calls that didn't match the signatures of their methods.  Once, I fixed that, it compiled with --aot.  Would have been better to have a more helpful error message.

**Comment by linuradu** [comment]
I had the following case:
with "ng build" worked fine. 

with "ng build -prod" failed with:
`ERROR in ..../src/$$_gendir/app/app.module.ngfactory.ts (499,57): Supplied parameters do not match any signature of call target.`

In app modules had:
```
providers: [
    {
      provide: Http,
      useFactory: translateHttpServiceLoader,
      deps: [XHRBackend, RequestOptions]
    }
  ]
```

and

```
export function translateHttpServiceLoader(backend: XHRBackend, options: RequestOptions, router: Router) {
  return new HttpService(backend, options, router);
}
```

I've found that `useFactory `should have only 2 params.
I've removed the last param and now is building fine also with "ng build -prod"

working version:
```
export function translateHttpServiceLoader(backend: XHRBackend, options: RequestOptions) {
  return new HttpService(backend, options);
}
```

**Comment by subatomicglue** [comment]
Just upgraded to latest angular-cli
```
@angular/cli: 1.2.7
node: 7.5.0
os: darwin x64
@angular/animations: 4.0.0
@angular/common: 4.0.0
@angular/compiler: 4.0.0
@angular/core: 4.0.0
@angular/forms: 4.0.0
@angular/http: 4.0.0
@angular/platform-browser: 4.0.0
@angular/platform-browser-dynamic: 4.0.0
@angular/router: 4.0.0
@angular/cli: 1.2.7
@angular/compiler-cli: 4.0.0
```

`ng build -prod` is giving me errors like this, which I do NOT get in `ng serve`: 

```
Property 'name' does not exist on type 'Object'.
Property 'editing' is private and only accessible within class 'AviaCommentComponent'.
```

I'm fine fixing these, but HOW do I get `ng serve` to inform us **during** development rather than **surprise us** later when building for production???  That is going to be a bad thing for us, i'd like to avoid

*Backstory:*

We used to use `1.0.0-beta30` of the `@angular/cli` and `2.3.1` of `@angular/core` and friends, For some reason this bug JUST showed up for us in the last two days after one of us did a `rm -rf node_modules && npm install` - the others on the team could still build with `-prod` fine...  Which prompted us to upgrade to Angular4 and latest 1.2.7 cli.  And we use all fixed versions in our package.json, so some sub dep must have changed to cause this because our versions are fixed (wtf indeed).

**Comment by subatomicglue** [comment]
You can replicate this problem with this sequence:

`ng new project`
`cd project`
edit `src/app/app.component.ts`  (put a private in front of `title:string;`) 

`npm start`   (works!   but should not...  very misleading during development)
> project@0.0.0 start /Users/username/src/project
> ng serve

** NG Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200 **
Hash: 7449fd53eff15e84d8b5                                                              
Time: 5212ms
chunk    {0} polyfills.bundle.js, polyfills.bundle.js.map (polyfills) 177 kB {4} [initial] [rendered]
chunk    {1} main.bundle.js, main.bundle.js.map (main) 5.28 kB {3} [initial] [rendered]
chunk    {2} styles.bundle.js, styles.bundle.js.map (styles) 10.5 kB {4} [initial] [rendered]
chunk    {3} vendor.bundle.js, vendor.bundle.js.map (vendor) 2.19 MB [initial] [rendered]
chunk    {4} inline.bundle.js, inline.bundle.js.map (inline) 0 bytes [entry] [rendered]
webpack: Compiled successfully.
```

`ng build -prod` **(doesn't work!)**
```
Hash: d066ff91b21cc6eafccb                                                              
Time: 7596ms
chunk    {0} polyfills.550cf10c9aa54b8194c7.bundle.js (polyfills) 177 kB {4} [initial] [rendered]
chunk    {1} main.2149231fc44e7510bd2f.bundle.js (main) 13.5 kB {3} [initial] [rendered]
chunk    {2} styles.d41d8cd98f00b204e980.bundle.css (styles) 69 bytes {4} [initial] [rendered]
chunk    {3} vendor.330bdb50b9e3457f8646.bundle.js (vendor) 849 kB [initial] [rendered]
chunk    {4} inline.95b082410a205220057a.bundle.js (inline) 0 bytes [entry] [rendered]

ERROR in /Users/username/src/project/src/$$_gendir/app/app.component.ngfactory.ts (57,31): Property 'title' is private and only accessible within class 'AppComponent'.
```

**Comment by subatomicglue** [comment]
From other threads, you can work around the problem.  Meaning, you CAN use `-prod` or `--aot` *without* changing any `private` to `public`, or fixing other errors...

Tested with my Angular `4.0.0` described above, which was installed by `cli@1.2.7`
Couple different ways to disable those errors in AoT mode:

Hack, which is not optimal and does not carry over on reinstalls.
```
npm install --save-dev @ngtools/webpack@1.2.14
rm -rf node_modules/\@angular/cli/node_modules/\@ngtools/webpack/
npm install --save-dev enhanced-resolve@3.3.0
```
and pin those versions.   Tested with cli@1.2.7.   `enhanced-resolve` fixes the error `Can't resolve './$$_gendir/app/app.module.ngfactory'`

OR

```
"devDependencies": {
 "@angular/cli": "1.0.0-rc.4",
 "@ngtools/webpack": "1.2.14",
  ...
}
```
without the `enhanced-resolve@3.3.0` from the first one or it wont work...


I tested both with a large project, and I no longer get those ~400+ errors about `private` and others when building with `ng build -prod` and `ng serve --aot`

References:
 - https://github.com/angular/angular-cli/issues/5623#issuecomment-289072266
 - https://github.com/angular/angular-cli/issues/4551#issuecomment-317459157

**Comment by subatomicglue** [comment]
So, Out of the box `@angular/cli@1.2.7` has an inconsistency between `ng serve` and `ng build -prod` and what errors stop the show. (see reproduction steps I posted above).

What is the long term solution going to be?  Anyone know?   

I see 3 possibilities:
- get `ng serve` to emit the same errors as `ng build -prod` so you're not surprised when making a prod build.
- modify `ng build -prod` so that these `private` errors are ignored or aren't errors (like it works in the workaround I posted above).
- Modify angular so that templates are compiled within the class scope, so html can reference `private` vars and make javascript style `object.key` references instead of `object['key']` - so it is then *correct* to reference private vars from within a template.

**Comment by wakawaka54** [comment]
Okay so how again are we suppose to figure out where the problem is if we have a:

ERROR in C:/.../src/$$_gendir/app/app.module.ngfactory.ts (159,52): Supplied parameters do not match any signature of call target.

??? Can we somehow keep the $$_gendir around? Is there some log file that it spits out? I have no clue what it is complaining about and I've spent about 5 hours looking through all the code and there is a lot of it.

**Comment by nathiyakumar** [comment]
i have error like this
ng -v
    _                      _                 ____ _     ___
   / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
  / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
 / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
/_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
               |___/
@angular/cli: 1.4.3
node: 7.10.1
os: linux x64
@angular/animations: error
@angular/common: error
@angular/compiler: 4.4.4
@angular/core: error
@angular/forms: error
@angular/http: error
@angular/platform-browser: error
@angular/platform-browser-dynamic: error
@angular/router: error
@angular/cli: error
@angular/compiler-cli: 4.4.4
@angular/language-service: error
typescript: 2.0.10

how to solve this error
package.json:

"dependencies": {
    "@angular/animations": "^4.0.0",
    "@angular/common": "^4.0.0",
    "@angular/compiler": "^4.0.0",
    "@angular/core": "^4.0.0",
    "@angular/forms": "^4.0.0",
    "@angular/http": "^4.0.0",
    "@angular/platform-browser": "^4.0.0",
    "@angular/platform-browser-dynamic": "^4.0.0",
    "@angular/router": "^4.0.0",
    "bootstrap": "^3.3.7",
    "core-js": "^2.4.1",
    "rxjs": "^5.1.0",
    "zone.js": "^0.8.4"
  },
  "devDependencies": {
    "@angular/cli": "^1.2.0",
    "@angular/compiler-cli": "^4.0.0",
    "@angular/language-service": "^4.0.0",
    "@types/jasmine": "2.5.45",
    "@types/node": "~6.0.60",
    "codelyzer": "~3.0.1",
    "jasmine-core": "~2.6.2",
    "jasmine-spec-reporter": "~4.1.0",
    "karma": "~1.7.0",
    "karma-chrome-launcher": "~2.1.1",
    "karma-cli": "~1.0.1",
    "karma-coverage-istanbul-reporter": "^1.2.1",
    "karma-jasmine": "~1.1.0",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "~5.1.2",
    "ts-node": "~3.0.4",
    "tslint": "~5.3.2",
    "typescript": "~2.0.0"
  }

**Comment by robrohan** [comment]
For anyone stuck on an old version of Angular and winds up here with this issue: we worked around it by adding a fake build step to our unit tests.  In our workflow, this will throw the error before a push, or at the worst throw the error on the test step before it gets all the way to the production build.

Something like this (package.json):

```
"scripts": {
   ...
    "fake-build": "ng build --prod --aot --target=production",
    ...
    "test": "npm run fake-build && ng test --browsers ChromeHeadless --single-run",
    ...
 },
```

This does make running the unit tests much slower, but at least stops the embarrassment of a surprise breaking of the production build.

**Comment by andalike** [comment]
The issue in my case was that My component was running, Stopped it and ran " ng build --prod --aot" and it worked.

**Comment by JahyrLopez** [comment]
help I have a problem problem mark me this error when I put the command 
ng build --prod

Date: 2019-06-26T19:47:29.429Z
Hash: 47649eb13e262d7eaef3
Time: 45551ms
chunk {0} runtime-es5.741402d1d47331ce975c.js (runtime) 1.41 kB [entry] [rendered]
chunk {1} main-es5.7c04bb8524a34ea08589.js (main) 828 kB [initial] [rendered]
chunk {2} polyfills-es5.b8d6dd293b73bbb8581d.js (polyfills) 132 kB [initial] [rendered]
chunk {scripts} scripts.000eefc49a6f20ab379c.js (scripts) 607 kB [entry] [rendered]

ERROR in scripts.000eefc49a6f20ab379c.js from Terser
Unexpected character '@' [scripts.000eefc49a6f20ab379c.js:17652,1]
